### ISSUE TITLE
Inconsistent behavior of RawKeyboardListener on different devices/Android versions

### ISSUE THREAD

**Issue by hicnar** [flutter/flutter#48296]
I am trying to get my flutter app to react to events generated when media keys are pressed on a little bluetooth keyboard paired with my android device.

Here's the code snipped that builds the widget I use to evaluate the scenario.

![raw-keyboard-listener-code](https://user-images.githubusercontent.com/1816194/71897913-dc836f00-314f-11ea-8bc9-30f23e1ec891.png)

If I run it on my Google Pixel C tablet running under Android 8.1.0, kernel version 3.18.0-g8ce8444 it works as expected, When the widget is displayed I can press the media buttons on the keyboard and the events are properly generated and passed to the app, handlers are fired etc.

On the other hand, if I test exactly the same code on my Samsung S9 running under Android 9, kernel version 4.9.59-17174122 my attempts to press the media buttons on the keyboard will take no effect until I press one of the numeric keys. After that, the following lines appear in the log 

![Screenshot from 2020-01-07 13:28:00](https://user-images.githubusercontent.com/1816194/71898579-9c24f080-3151-11ea-988c-ac8e52178627.png)

After that the media keys on the keyboard work as expected.
Is there any quick workaround to that problem?

Finally a question regarding iOS. I am aware that RawKeyboardListerer is not implemented for iOS, but is there a way for a flutter app running on iOS device to react to media buttons like in the above scenario?

**Comment by VladyslavBondarenko** [comment]
Hi @hicnar 
Could you please provide your code sample and your logs in the text form instead of posting images?
About your second question, you may want to follow [#31945](https://github.com/flutter/flutter/issues/31945)
Thank you

**Comment by hicnar** [comment]
Thank you for your response Vladyslav

Right below is the code snippet I used for testing

```
Widget _buildAndroid(BuildContext context) {
    return Container(
        child: ListView(children: [
      SizedBox(height: 10),
      RawKeyboardListener(
          focusNode: _focusNode,
          onKey: _handleKeyEvent,
          autofocus: true,
          child: Ink(
              child: ListTile(
                  onTap: () {
                    setState(() {
                      _focusNode.requestFocus();
                    });
                  },
                  leading: Text(_playPause),
                  trailing: Icon(Icons.play_arrow, color: Colors.blue)))),
      SizedBox(height: 5),
      RawKeyboardListener(
          focusNode: _focusNode2,
          onKey: _handleKeyEvent,
          child: Ink(
              child: ListTile(
                  onTap: () {
                    setState(() {
                      _focusNode2.requestFocus();
                    });
                  },
                  leading: Text(_scrollUp),
                  trailing: Icon(Icons.fast_forward, color: Colors.blue)))),
      SizedBox(height: 5),
      Ink(
          child: ListTile(
              onTap: () {},
              leading: Text(_scrollDown),
              trailing: Icon(Icons.fast_rewind, color: Colors.blue))),
      SizedBox(height: 5),
      ListTile(
          leading: Text(_increaseScrollSpeed),
          trailing: Icon(Icons.add_circle_outline, color: Colors.blue)),
      SizedBox(height: 5),
      ListTile(
          leading: Text(_decreaseScrollSpeed),
          trailing: Icon(Icons.remove_circle_outline, color: Colors.blue)),
      SizedBox(height: 10)
    ]));
  }
```

And here's the log excerpt showing what's printed out right after a non media key is pressed that in a way "unlocks" the functionality so it starts to work on my S9 just as it does on the Pixel C (those lines are not printed when I test the app on my Pixel C).

```
D/InputMethodManager(24337): prepareNavigationBarInfo() DecorView@9037b9c[MainActivity]
D/InputMethodManager(24337): getNavigationBarColor() -855310
V/InputMethodManager(24337): Starting input: tba=com.test.my_app ic=null mNaviBarColor -855310 mIsGetNaviBarColorSuccess true , NavVisible : true , NavTrans : false
D/InputMethodManager(24337): startInputInner - Id : 0
I/InputMethodManager(24337): startInputInner - mService.startInputOrWindowGainedFocus
D/InputTransport(24337): Input channel constructed: fd=104
D/InputTransport(24337): Input channel destroyed: fd=102
D/ViewRootImpl@90b9f0f[MainActivity](24337): ViewPostIme key 1
```

I am aware of #31945 actually I am one of those who gave it a thumb up. Do you reckon it is possible to write a flutter plugin that handles that in iOS - at least the media buttons?

**Comment by hicnar** [comment]
@VladyslavBondarenko @goderbauer 

Hi folks, 

Sorry to bother you, but did you or any of the flutter members team have a chance to look into that? 
It may happen that this problem will soon become a show stopper inhibiting the release of my project.

Thanks in advance!
K.

**Comment by hicnar** [comment]
There is a hint that may give you some idea about what may potentially be wrong. The functionality begins to work as soon as any of the text fields in the application (even in a completely different part of the application) gets in focus.

**Comment by darshankawar** [comment]
Hi @hicnar 
Is this still an issue ?
If so, can you provide `flutter doctor -v` and steps to test the code sample you shared earlier ? because you mentioned to have used a bluetooth enabled keyboard paired with your android device.
Thanks.

**Comment by hicnar** [comment]
@darshankawar Thank you for your response. I will be re-testing this functionality over this weekend and in case the problem persists I will provide a response along with a full code sample (including the main()) so that you can simply copy-paste and run the test on your device.

**Comment by hicnar** [comment]
@darshankawar 

Below is a simple app that can be used to see the inconsistency in the behaviour.

```
import 'package:flutter/material.dart';

void main() => runApp(ExampleApp());

class ExampleApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Bluetooth Remote Media Keys Test',
      theme: new ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: ContentTab(),
    );
  }
}


class ContentTab extends StatefulWidget {

  @override
  _ContentTabState createState() => _ContentTabState();
}

class _ContentTabState extends State<ContentTab> {

  final _focusNode = FocusNode();
  final TextEditingController controller = TextEditingController();

  @override
  Widget build(context) => _buildAndroid(context);

  Widget _buildAndroid(BuildContext context) {
    return Scaffold(
      body: Center(
      child: Container(
        child: ListView(
          children: [
            TextField(
              controller: controller,
            ),
            RawKeyboardListener(
              focusNode: _focusNode,
              onKey: (event) {
                controller.text = event.data.physicalKey.debugName;
                debugPrint('--------------- received keyboard event: $event');
              },
              autofocus: true,
              child: Ink(
                color: Colors.blue,
                child: ListTile(
                  onTap: () {
                    setState(() {
                      _focusNode.requestFocus();
                    });
                  },
                  leading: Text('Tap to manually request focus'),
                )
              )
            )
          ]
        )
    )));
  }
}
```
I have run my tests on two devices, Google Pixel C tablet and a Samsung S9 mobile phone.
If I run it on Pixel C pa

Steps to reproduce:
Pair a remote presentation keyboard with your mobile. 
I am using one of the Satechi keyboards
![bluetooth-multi-media-remote-remotes-satechi-435448_1024x](https://user-images.githubusercontent.com/1816194/93838324-2336dc00-fc81-11ea-8ca6-77656ad41a8a.jpg) 

But the brand not important, what is important is that the keyboard you use has keys like: "Media Play Pause", "Media Track Next", "Media Track Previous", "Audio Volume Up", "Audio Volume Down" etc.

Once your keyboard is paired with your Android device run the application and try to press some of the media keys. What should happen and pretty much happens when I run it on Pixel C, once I press any media key I see its debug name displayed in the TextField.

It works this way only on Pixel C though. On S9 in order for this to work you will need to tap on the TextField so that it gains focus and then tap again on the blue area beneath the TextField (labeled "Tap to manually request focus") so that the focus is transferred to the FocusNode associated with the RawKeyboardListener and only then pressing the media keys on your remote will result in receiving the events that after being handled will print the debug names of the keys pressed in the TextField.

Hope this helps!

**Comment by darshankawar** [comment]
Thanks @hicnar for the detailed report and steps to replicate.
Since I don't have any external bluetooth device to replicate this issue, I am keeping this open for further analysis, as you mentioned the issue is still occurring.

**Comment by Renzo-Olivares** [comment]
cc from text-input triage @dkwingsmt @gspencergoog did you have any thoughts on this issue? I know there is `KeyboardListener` available now, is that preferred over RawKeyboardListener?

**Comment by gspencergoog** [comment]
Yes, `KeyboardListener` is preferred over `RawKeyboardListener`, but I'm not sure that will have any effect on this problem.

This seems like the cause is something lower down in the stack, since the different Android distributions produce different results. It could be that Samsung just doesn't produce the same key events at a low level.  To debug this, we'd need to get the devices and see what it is producing when those keys are pressed.

It's interesting that the behavior changes when you focus a text field. The text field does open up an [InputConnection](https://developer.android.com/reference/android/view/inputmethod/InputConnection) when focused, so maybe that is what is turning it on on Samsung, but it's always on for Pixel?
